DistroLauncher/named_mutex.h

```diff
@@ -0,0 +1,219 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+#include "stdafx.h"
+
+/*
+ *  Abstract
+ *  --------
+ *  A RAII class to deal with Windows inter-process synchronization.
+ *
+ *  Usage
+ *  --------
+ *  Best is to create a single NamedMutex as a static variable. If you believe
+ *  it's not going to be used in most executions, then use
+ *  ```
+ *  static my_mutex = NamedMutex("unique_identifyier", true);
+ *  ```
+ *  to enable lazy initialization. Otherwise, use
+ *  ```
+ *  static my_mutex = NamedMutex("unique_identifyier");
+ *  ```
+ *
+ *  In order to acquire a lock, you have two options.
+ *  Use the first one unless you've got a good reason not to.
+ *   1. Use the monadic interface:
+ *      ```
+ *      my_mutex.lock().and_then([](){
+ *          // do stuff
+ *      }).or_else([](){
+ *          throw "Oh, no!";
+ *      });
+ *      ```
+ *      Note that the ordering of the functions doesn't matter,
+ *      and both are optional and repeatable, so long as no
+ *      exceptions are thrown. In this case, the mutex will be
+ *      released and the exception re-thrown.
+ *
+ *   2. Use a scope guard:
+ *      ```
+ *      NamedMutex::Guard scope_guard = my_mutex.lock();
+ *      if(scope_guard.ok()) {
+ *          // do stuff
+ *      } else {
+ *          throw "Oh, no!";
+ *      }
+ *      ```
+ *      Be careful! If your program throws and it's not caught,
+ *      your mutex may not be released. Option 1 takes care of
+ *      this automatically, hence why it's recommended.
+ */
+
+// Templated for testing/dependency injection reasons, always use the alias at the end of the file
+template <typename MutexAPI> class NamedMutexWrapper
+{
+  public:
+    NamedMutexWrapper() = delete;
+    NamedMutexWrapper(NamedMutexWrapper&) = delete;
+    NamedMutexWrapper(NamedMutexWrapper&&) = delete;
+
+    /// Some mutexes are almost never used. This makes for an easy optimization by not initializing them unless needed.
+    NamedMutexWrapper(std::wstring name, bool lazy_init = false) :
+        mutex_handle(nullptr), mutex_name(L"WSL_" + DistributionInfo::Name + L"_" + name)
+    {
+        if (!lazy_init) {
+            create();
+        }
+    }
+
+    virtual ~NamedMutexWrapper() noexcept
+    {
+        destroy();
+    }
+
+    class Lock
+    {
+      public:
+        constexpr Lock() noexcept : parent_(nullptr), response_(0)
+        {
+        }
+
+        Lock(NamedMutexWrapper& parent) noexcept : parent_(&parent), response_(parent.wait_and_acquire())
```

Comment:
I'm a bit surprised clang-tidy didn't complain about the missing `explicit` keyword here.